Favicon

jmglov opened the pull request:
Please answer the following questions and leave the below in as part of your PR.

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/babashka/babashka/blob/master/doc/dev.md#start-with-an-issue-before-writing-code)

- [x] I have updated the [CHANGELOG.md](https://github.com/babashka/babashka/blob/master/CHANGELOG.md) file with a description of the addressed issue.

borkdude commented:
This PR breaks caching. I'm seeing all `.md` pages being re-rendered on every `bb render` call.

jmglov commented:
Odd. I’ll look at it little later. I guess it might be time to start
thinking about how to have automated tests for some of this stuff. 😭

On Mon, 18 Jul 2022 at 13:52, Michiel Borkent ***@***.***>
wrote:

> This PR breaks caching. I'm seeing all .md pages being re-rendered on
> every bb render call.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/borkdude/quickblog/pull/9#issuecomment-1187195963>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWWRRU6XVI3TPUWWRD56TVUVAOHANCNFSM533R7C2A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Sent from my iPhone. Please excuse the brevity and potential typos.

jmglov commented:
You can follow the instructions in the README or look at my
quickblog-branding branch, which adds one. If it’s working, you’ll see a
little icon in your tab bar.

On Mon, 18 Jul 2022 at 13:57, Michiel Borkent ***@***.***>
wrote:

> How can I verify that this works, locally?
>
> [image: Screenshot 2022-07-18 at 13 56 09]
> <https://user-images.githubusercontent.com/284934/179506232-981824a7-1f20-4dc7-8652-9c2afad8f7c1.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/borkdude/quickblog/pull/9#issuecomment-1187207697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWWRWZ37RZQ6MJRIMBYGDVUVBBFANCNFSM533R7C2A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Sent from my iPhone. Please excuse the brevity and potential typos.

jmglov commented:
@borkdude Everything is fixed now. :)

jmglov commented:
@borkdude The default is possible, but let's do that in my rebranding PR, which adds a default favicon to the blog. I'll rename the argument later tonight.

borkdude commented:
The reason I'm asking for a default is so I can test it locally. Now I have no clue if things are working without creating a favicon thing myself.

jmglov commented:
OK, I'll add the default in this PR, then. I'll wait until the morning to do it, though. ;)

jmglov commented:
@borkdude Added the default.

borkdude commented:
I pulled your branch, did bb clean and rm -rf templates, re-built my blog with `:enable-favicon true` but I see no favicon. Am I doing it wrong?

jmglov commented:
The option is called `:favicon` now, as per your suggestion. 😅

borkdude commented:
I retried with favicon true and cleared out everything, but still don't see it. Can you try with my blog, to verify?

jmglov commented:
Works for me:
```
[jmglov@laurana:~/Documents/code/clojure/borkdude-blog]$ bb clean && rm -rf templates/ && bb render --favicon true
Writing default resource: templates/style.css
Writing public/assets/boot-lein-2015/hacker-news-lein2boot.png
Writing public/assets/browserconfig.xml
Writing public/assets/mstile-150x150.png
Writing public/assets/safari-pinned-tab.svg
Writing public/assets/site.webmanifest
Writing public/assets/android-chrome-512x512.png
Writing public/assets/favicon-32x32.png
Writing public/assets/oredev-2014/ryan-pedestal.jpg
Writing public/assets/oredev-2014/vegan-food-oredev.jpg
Writing public/assets/oredev-2014/badge.jpg
Writing public/assets/oredev-2014/michiel_clojurescript_ftw.jpg
Writing public/assets/apple-touch-icon.png
Writing public/assets/favicon-16x16.png
Writing public/assets/android-chrome-192x192.png
Writing public/assets/favicon.ico
[...]
[jmglov@laurana:~/Documents/code/clojure/borkdude-blog]$ grep -A 10 'Features > favicon' public/index.html 
    <!-- See Features > favicon in README.md -->
    <link rel="apple-touch-icon" sizes="180x180" href="assets/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="assets/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="assets/favicon-16x16.png">
    <link rel="manifest" href="assets/site.webmanifest">
    <link rel="mask-icon" href="assets/safari-pinned-tab.svg" color="#5bbad5">
    <meta name="msapplication-TileColor" content="#da532c">
    <meta name="theme-color" content="#ffffff">

  </head>
  <body>

```

And I'm also seeing the icon with my eyes. This is so odd.

Can you run the same commands on your blog and see what happens?

jmglov commented:
It works for me. My guess is that you ran `bb render --favicon true` but then didn't add the `--favicon true` to the `bb watch` command, maybe?

borkdude commented:
I added `:favicon true` to my config in bb.edn.

jmglov commented:
Incredibly odd. OK, let me try that.

jmglov commented:
Works for me with your blog:

![Screenshot_2022-07-19_13-53-28](https://user-images.githubusercontent.com/224070/179745175-9a8f93f3-8f69-4d02-a143-984975f5f7f1.png)

borkdude commented:
Here is what I see:

<img width="982" alt="Screenshot 2022-07-19 at 14 03 15" src="https://user-images.githubusercontent.com/284934/179745740-129cb587-1921-4655-a60d-7dfce784da3a.png">
<img width="691" alt="Screenshot 2022-07-19 at 14 03 25" src="https://user-images.githubusercontent.com/284934/179745757-6ada54c2-1f65-48c3-98d1-9202c6da0069.png">
<img width="946" alt="Screenshot 2022-07-19 at 14 03 36" src="https://user-images.githubusercontent.com/284934/179745762-6970cd0f-930b-460e-b539-a875bd9b4701.png">

borkdude commented:
Network tab:

<img width="640" alt="Screenshot 2022-07-19 at 14 06 31" src="https://user-images.githubusercontent.com/284934/179746296-dc1beb67-6686-4935-a04f-d0b0adc70b9a.png">

borkdude commented:
In Safari it does seem to work:

<img width="605" alt="Screenshot 2022-07-19 at 14 07 39" src="https://user-images.githubusercontent.com/284934/179746568-a5178e4c-df8f-4af5-b8f5-48b0f9f86fd3.png">

Just not with Chrome.app.

jmglov commented:
The favicon generator I used recommended that you put the favicon assets at the root of your website, so maybe Chrome.app only supports that. Does it try to load `favicon.ico`, by any chance?

Oddly enough, Chrome on Linux does work for me. So does Firefox.

I'm not really sure where to go from here. Do you have any ideas or suggestions?

borkdude commented:
@jmglov Since it works in Safari, I think it's sufficient. If someone wants this to work in Chrome.app they can submit another PR.

Two things remaining:

1) When I edit a blog post, it re-renders:
 
Writing page: public/archive.html
Writing page: public/index.html

Why is that?

2) The "meter" icon: are there any other good candidates for a default icon? Or was this part of your "re-branding"?

borkdude commented:
> Neither the index or archive page currently have any caching happening.

Sure, but why are they re-rendered? On main this isn't happening, so this introduced a minor regression which I'd like to avoid.

borkdude commented:
@jmglov OK, let's go with this minimal one: https://thenounproject.com/icon/writing-62338/

jmglov commented:
I'll update to that icon. I didn't realise there was a regression with the rendering. I'll fix that as well, but I probably won't have the PR updated in until later tonight.

jmglov commented:
@borkdude Icon switched and caching added. :)

borkdude commented:
@jmglov When I clean out my assets folder's previously copied favicon, and then do bb render, I get:

```
Writing public/assets/favicon-16x16.png
----- Error --------------------------------------------------------------------
Type:     java.nio.file.FileAlreadyExistsException
Message:  public/assets/favicon-16x16.png
```

Another thought: can we put the favicon assets into `assets/favicon/`? This will make it easier to clear it and get the default favicon back, while keeping your own assets.

jmglov commented:
I know what's causing the `FileAlreadyExistsException` and will fix. I agree about moving the favicon assets as well. Will update the PR today or tomorrow. (Enjoying the ☀️ in Malaga atm!)

borkdude commented:
@jmglov I tried the recent commits, seems to work now. But I'm still seeing the meter icon

jmglov commented:
Have you run `bb clean`, removed your templates directory, and deleted the old assets? My guess is that some pages haven't been re-rendered and are still pointing to the old icons. You may also need to do a Shift-refresh in your browser in case the favicon is being cached.

borkdude commented:
That was it! Merging.

jmglov commented:
🎉

borkdude commented:
@jmglov Speaking of caching: I think one thing that slipped through the cracks is this:

Whenever I update a blog post, the index page should also be updated, since the text there will change. Not sure why this isn't happening anymore.

borkdude commented:
I'll make an issue for that.

jmglov commented:
OK, cool. That will definitely be fixed in the new caching branch.
